# Hands-on Activity 1.1 | Optimization and Knapsack Problem

#### Objective(s):

This activity aims to demonstrate how to apply  greedy and brute force algorithms to solve optimization problems

#### Intended Learning Outcomes (ILOs):
* Demonstrate how to solve knapsacks problems using greedy algorithm
* Demonstrate how to  solve knapsacks problems using brute force algorithm


#### Resources:
* Jupyter Notebook


#### Procedures:

1. Create a Food class that defines the following:
* name of the food
* value of the food
* calories of the food

2. Create the following methods inside the Food class:
* A method that returns the value of the food
* A method that returns the cost of the food
* A method that calculates the density of the food (Value / Cost)
* A method that returns a string to display the name, value and calories of the food

In [23]:
class Food(object):
    def __init__(self, n, v, w):
        # Make the variables private
        self.name = n
        self.value = v
        self.calories = w
    def getValue(self):
        return self.value
    def getCost(self):
        return self.calories
    def density(self):
        return self.getValue()/self.getCost()
    def __str__(self):
        return self.name + ': <' + str(self.value)+ ', ' + str(self.calories) + '>'

3. Create a buildMenu method that builds the name, value and calories of the food


In [24]:
def buildMenu(names, values, calories):
    menu = []
    for i in range(len(values)):
        menu.append(Food(names[i], values[i],calories[i]))
    return menu

4. Create a method greedy to return total value and cost of added food based on the desired maximum cost

In [25]:
def greedy(items, maxCost, keyFunction):
    """Assumes items a list, maxCost >= 0,         keyFunction maps elements of items to numbers"""
    itemsCopy = sorted(items, key = keyFunction,
                       reverse = True)
    result = []
    totalValue, totalCost = 0.0, 0.0
    for i in range(len(itemsCopy)):
        if (totalCost+itemsCopy[i].getCost()) <= maxCost:
            result.append(itemsCopy[i])
            totalCost += itemsCopy[i].getCost()
            totalValue += itemsCopy[i].getValue()
    return (result, totalValue)

5. Create a testGreedy method to test the greedy method

In [26]:
def testGreedy(items, constraint, keyFunction):
    taken, val = greedy(items, constraint, keyFunction)
    print('Total value of items taken =', val)
    for item in taken:
        print('   ', item)

In [27]:
def testGreedys(foods, maxUnits):
    print('Use greedy by value to allocate', maxUnits,          'calories')
    testGreedy(foods, maxUnits, Food.getValue)
    print('\nUse greedy by cost to allocate', maxUnits,          'calories')
    testGreedy(foods, maxUnits, lambda x: 1/Food.getCost(x))
    print('\nUse greedy by density to allocate', maxUnits,          'calories')
    testGreedy(foods, maxUnits, Food.density)

6. Create arrays of food name, values and calories
7. Call the buildMenu to create menu for food
8. Use testGreedys method to pick food according to the desired calories

In [28]:
names = ['wine', 'beer', 'pizza', 'burger', 'fries','cola', 'apple', 'donut', 'cake']
values = [89,90,95,100,90,79,50,10]
calories = [123,154,258,354,365,150,95,195]
foods = buildMenu(names, values, calories)
testGreedys(foods, 2000)

Use greedy by value to allocate 2000 calories
Total value of items taken = 603.0
    burger: <100, 354>
    pizza: <95, 258>
    beer: <90, 154>
    fries: <90, 365>
    wine: <89, 123>
    cola: <79, 150>
    apple: <50, 95>
    donut: <10, 195>

Use greedy by cost to allocate 2000 calories
Total value of items taken = 603.0
    apple: <50, 95>
    wine: <89, 123>
    cola: <79, 150>
    beer: <90, 154>
    donut: <10, 195>
    pizza: <95, 258>
    burger: <100, 354>
    fries: <90, 365>

Use greedy by density to allocate 2000 calories
Total value of items taken = 603.0
    wine: <89, 123>
    beer: <90, 154>
    cola: <79, 150>
    apple: <50, 95>
    pizza: <95, 258>
    burger: <100, 354>
    fries: <90, 365>
    donut: <10, 195>


Task 1: Change the maxUnits to 100

In [29]:
names = ['wine', 'beer', 'pizza', 'burger', 'fries','cola', 'apple', 'donut', 'cake']
values = [89,90,95,100,90,79,50,10]
calories = [123,154,258,354,365,150,95,195]
foods = buildMenu(names, values, calories)
testGreedys(foods, 100)

Use greedy by value to allocate 100 calories
Total value of items taken = 50.0
    apple: <50, 95>

Use greedy by cost to allocate 100 calories
Total value of items taken = 50.0
    apple: <50, 95>

Use greedy by density to allocate 100 calories
Total value of items taken = 50.0
    apple: <50, 95>


Task 2: Modify codes to add additional weight (criterion) to select food items.

In [30]:
class Food(object):
    def __init__(self, n, v, w, wt):
        # Make the variables private
        self.name = n
        self.value = v
        self.calories = w
        self.weight = wt #new attribute

    def getValue(self):
        return self.value

    def getCost(self):
        return self.calories

    def getWeight(self): #new method
        return self.weight

    def density(self):
        return self.getValue() / self.getCost()

    def weightDensity(self): #new method
        return self.getValue() / self.getWeight()

    def __str__(self):
        return self.name + ': <' + str(self.value) + ', ' + str(self.calories) + ', ' + str(self.weight) + '>'

def buildMenu(names, values, calories, weights):
    menu = []
    for i in range(len(values)):
        menu.append(Food(names[i], values[i], calories[i], weights[i])) #added weights[i]
    return menu
def testGreedy(items, constraint, keyFunction):
    taken, val = greedy(items, constraint, keyFunction)
    print('Total value of items taken =', val)
    for item in taken:
        print('   ', item)

Task 3: Test your modified code to test the greedy algorithm to select food items with your additional weight.

In [31]:
def testGreedys(foods, maxUnits):
    print('Use greedy by value to allocate', maxUnits, 'calories')
    testGreedy(foods, maxUnits, Food.getValue)
    print('\nUse greedy by cost to allocate', maxUnits, 'calories')
    testGreedy(foods, maxUnits, lambda x: 1 / Food.getCost(x))
    print('\nUse greedy by density to allocate', maxUnits, 'calories')
    testGreedy(foods, maxUnits, Food.density)
    print('\nUse greedy by weight to allocate', maxUnits, 'calories') #new test
    testGreedy(foods, maxUnits, Food.getWeight)
    print('\nUse greedy by weight density to allocate', maxUnits, 'calories')#new test
    testGreedy(foods, maxUnits, Food.weightDensity)

names = ['wine', 'beer', 'pizza', 'burger', 'fries', 'cola', 'apple', 'donut', 'cake']
values = [89, 90, 95, 100, 90, 79, 50, 10]
calories = [123, 154, 258, 354, 365, 150, 95, 195]
weights = [1, 1.5, 2, 2.5, 3, 1.2, 0.5, 0.3]  # Example weights
foods = buildMenu(names, values, calories, weights)
testGreedys(foods, 100)  # Changed maxUnits from 2000 to 100

Use greedy by value to allocate 100 calories
Total value of items taken = 50.0
    apple: <50, 95, 0.5>

Use greedy by cost to allocate 100 calories
Total value of items taken = 50.0
    apple: <50, 95, 0.5>

Use greedy by density to allocate 100 calories
Total value of items taken = 50.0
    apple: <50, 95, 0.5>

Use greedy by weight to allocate 100 calories
Total value of items taken = 50.0
    apple: <50, 95, 0.5>

Use greedy by weight density to allocate 100 calories
Total value of items taken = 50.0
    apple: <50, 95, 0.5>


Using 1000 instead of 100

In [33]:
def testGreedys(foods, maxUnits):
    print('Use greedy by value to allocate', maxUnits, 'calories')
    testGreedy(foods, maxUnits, Food.getValue)
    print('\nUse greedy by cost to allocate', maxUnits, 'calories')
    testGreedy(foods, maxUnits, lambda x: 1 / Food.getCost(x))
    print('\nUse greedy by density to allocate', maxUnits, 'calories')
    testGreedy(foods, maxUnits, Food.density)
    print('\nUse greedy by weight to allocate', maxUnits, 'calories') #new test
    testGreedy(foods, maxUnits, Food.getWeight)
    print('\nUse greedy by weight density to allocate', maxUnits, 'calories')#new test
    testGreedy(foods, maxUnits, Food.weightDensity)

names = ['wine', 'beer', 'pizza', 'burger', 'fries', 'cola', 'apple', 'donut', 'cake']
values = [89, 90, 95, 100, 90, 79, 50, 10]
calories = [123, 154, 258, 354, 365, 150, 95, 195]
weights = [1, 1.5, 2, 2.5, 3, 1.2, 0.5, 0.3]  # Example weights
foods = buildMenu(names, values, calories, weights)
testGreedys(foods, 1000)  # Changed maxUnits from 2000 to 100

Use greedy by value to allocate 1000 calories
Total value of items taken = 424.0
    burger: <100, 354, 2.5>
    pizza: <95, 258, 2>
    beer: <90, 154, 1.5>
    wine: <89, 123, 1>
    apple: <50, 95, 0.5>

Use greedy by cost to allocate 1000 calories
Total value of items taken = 413.0
    apple: <50, 95, 0.5>
    wine: <89, 123, 1>
    cola: <79, 150, 1.2>
    beer: <90, 154, 1.5>
    donut: <10, 195, 0.3>
    pizza: <95, 258, 2>

Use greedy by density to allocate 1000 calories
Total value of items taken = 413.0
    wine: <89, 123, 1>
    beer: <90, 154, 1.5>
    cola: <79, 150, 1.2>
    apple: <50, 95, 0.5>
    pizza: <95, 258, 2>
    donut: <10, 195, 0.3>

Use greedy by weight to allocate 1000 calories
Total value of items taken = 285.0
    fries: <90, 365, 3>
    burger: <100, 354, 2.5>
    pizza: <95, 258, 2>

Use greedy by weight density to allocate 1000 calories
Total value of items taken = 413.0
    apple: <50, 95, 0.5>
    wine: <89, 123, 1>
    cola: <79, 150, 1.2>
    beer: 

9. Create method to use  Bruteforce algorithm instead of greedy algorithm

In [34]:
def maxVal(toConsider, avail):
    """Assumes toConsider a list of items, avail a weight
       Returns a tuple of the total value of a solution to the
         0/1 knapsack problem and the items of that solution"""
    if toConsider == [] or avail == 0:
        result = (0, ())
    elif toConsider[0].getCost() > avail:
        #Explore right branch only
        result = maxVal(toConsider[1:], avail)
    else:
        nextItem = toConsider[0]
        #Explore left branch
        withVal, withToTake = maxVal(toConsider[1:],
                                     avail - nextItem.getCost())
        withVal += nextItem.getValue()
        #Explore right branch
        withoutVal, withoutToTake = maxVal(toConsider[1:], avail)
        #Choose better branch
        if withVal > withoutVal:
            result = (withVal, withToTake + (nextItem,))
        else:
            result = (withoutVal, withoutToTake)
    return result

In [35]:
def testMaxVal(foods, maxUnits, printItems = True):
    print('Use search tree to allocate', maxUnits,
          'calories')
    val, taken = maxVal(foods, maxUnits)
    print('Total costs of foods taken =', val)
    if printItems:
        for item in taken:
            print('   ', item)

In [36]:
names = ['wine', 'beer', 'pizza', 'burger', 'fries','cola', 'apple', 'donut', 'cake']
values = [89,90,95,100,90,79,50,10]
calories = [123,154,258,354,365,150,95,195]
foods = buildMenu(names, values, calories)
testMaxVal(foods, 2400)

TypeError: buildMenu() missing 1 required positional argument: 'weights'

Added weights

In [ ]:
names = ['wine', 'beer', 'pizza', 'burger', 'fries','cola', 'apple', 'donut', 'cake']
values = [89,90,95,100,90,79,50,10]
calories = [123,154,258,354,365,150,95,195]
foods = buildMenu(names, values, calories,weights)
testMaxVal(foods, 2400)

Use search tree to allocate 2400 calories
Total costs of foods taken = 603
    donut: <10, 195, 0.3>
    apple: <50, 95, 0.5>
    cola: <79, 150, 1.2>
    fries: <90, 365, 3>
    burger: <100, 354, 2.5>
    pizza: <95, 258, 2>
    beer: <90, 154, 1.5>
    wine: <89, 123, 1>


#### Supplementary Activity:

Real-world Problem: (Oversimplified) Creating the best single season NBA team 2024-2025

Constraints: Cap Space = $141000000

Maximize: 2K rating

The goal is to create a maximum of 15-man roster that is theoretically the best team ever assembled right now with Cap Space to consider.

Limitations: There are over 500 players in the NBA and I could not find a database that stores all of them. I sourced my data based on this website:

The maximum player rating is 99. The best player in the NBA right now is 98.

https://www.2kratings.com/lists/top-100-highest-paid-nba-players

I listed 73 players only.



In [19]:
class Player(object):
    def __init__(self, n, r, s):
        # Make the variables private
        self.__name = n
        self.__rating = r
        self.__salary = s

    def getName(self):
        return self.__name

    def getRating(self):
        return self.__rating

    def getSalary(self):
        return self.__salary

    def __str__(self):
        return f"{self.__name}: <{self.__rating}, {self.__salary}>"

def buildList(names, ratings, salaries):
    player_list = []
    for i in range(len(ratings)):
        player_list.append(Player(names[i], ratings[i], salaries[i]))
    return player_list

##############################################################################
def greedy(items, maxCost, keyFunction):
    """Assumes items a list, maxCost >= 0, keyFunction maps elements of items to numbers"""
    itemsCopy = sorted(items, key=keyFunction, reverse=True)
    result = []
    totalRating, totalCost = 0.0, 0.0
    for i in range(len(itemsCopy)):
        if (totalCost + itemsCopy[i].getSalary()) <= maxCost:
            result.append(itemsCopy[i])
            totalCost += itemsCopy[i].getSalary()
            totalRating += itemsCopy[i].getRating()
    return (result, totalRating)

def testGreedy(items, constraint, keyFunction):
    taken, val = greedy(items, constraint, keyFunction)
    print('Total rating of players taken =', val)
    for item in taken:
        print('   ', item)

def testGreedys(players, maxUnits):
    print('Use greedy by rating to allocate', maxUnits, 'capspace')
    testGreedy(players, maxUnits, Player.getRating)

    print('Use greedy by salary to allocate', maxUnits, 'capspace')
    testGreedy(players, maxUnits, lambda x: -x.getSalary())
##############################################################################
##############################################################################
def maxVal(toConsider, avail):
    """Assumes toConsider a list of items, avail a salary cap
       Returns a tuple of the total rating of a solution to the
       0/1 knapsack problem and the items of that solution"""
    if toConsider == [] or avail == 0:
        result = (0, ())
    elif toConsider[0].getSalary() > avail:
        # Explore right branch only
        result = maxVal(toConsider[1:], avail)
    else:
        nextPlayer = toConsider[0]
        # Explore left branch
        withVal, withToTake = maxVal(toConsider[1:], avail - nextPlayer.getSalary())
        withVal += nextPlayer.getRating()
        # Explore right branch
        withoutVal, withoutToTake = maxVal(toConsider[1:], avail)
        # Choose better branch
        if withVal > withoutVal:
            result = (withVal, withToTake + (nextPlayer,))
        else:
            result = (withoutVal, withoutToTake)
    return result

def testMaxVal(players, maxSalary, printItems=True):
    print('(BRUTE FORCE) Use search tree to allocate players within a salary cap of', maxSalary)
    val, taken = maxVal(players, maxSalary)
    print('Total rating of players taken =', val)
    if printItems:
        for player in taken:
            print('   ', player)
##############################################################################
names = [
    "Stephen Curry", "Kevin Durant", "Joel Embiid", "LeBron James", "Nikola Jokic",
    "Bradley Beal", "Paul George", "Damian Lillard", "Giannis Antetokounmpo", "Kawhi Leonard",
    "Jimmy Butler", "Klay Thompson", "Rudy Gobert", "Fred VanVleet", "Anthony Davis",
    "Trae Young", "Zach LaVine", "Luka Doncic", "Tobias Harris", "Ben Simmons",
    "Pascal Siakam", "Kyrie Irving", "Jrue Holiday", "Devin Booker", "Karl-Anthony Towns",
    "Kristaps Porzingis", "C.J. McCollum", "James Harden", "Zion Williamson", "Ja Morant",
    "Darius Garland", "Brandon Ingram", "Jamal Murray", "Shai Gilgeous-Alexander", "Michael Porter Jr.",
    "Donovan Mitchell", "Jayson Tatum", "De’Aaron Fox", "Bam Adebayo", "Deandre Ayton",
    "Jaylen Brown", "Gordon Hayward", "Chris Paul", "Domantas Sabonis", "Kyle Lowry",
    "Khris Middleton", "DeMar DeRozan", "Julius Randle", "Jordan Poole", "Jerami Grant",
    "Jaren Jackson Jr.", "Tyler Herro", "Jalen Brunson", "Cameron Johnson", "Kyle Kuzma",
    "John Collins", "Brook Lopez", "Mike Conley", "Andrew Wiggins", "Anfernee Simons",
    "R.J. Barrett", "Jordan Clarkson", "Terry Rozier III", "Dillon Brooks", "Malcolm Brogdon",
    "Draymond Green", "Aaron Gordon", "Bruce Brown", "Mikal Bridges", "Myles Turner",
    "Clint Capela", "Lonzo Ball", "Spencer Dinwiddie",
]

ratings = [
    95, 96, 98, 96, 98, 86, 89, 89, 97, 92, 93, 82, 87, 85, 95, 89, 83, 97, 81, 77,
    90, 94, 87, 95, 88, 88, 85, 85, 88, 93, 82, 86, 89, 97, 84, 93, 96, 89, 88, 85,
    93, 76, 79, 90, 78, 87, 88, 86, 80, 82, 87, 83, 94, 79, 83, 81, 81, 81, 81, 79,
    83, 79, 81, 80, 81, 83, 85, 77, 83, 85, 81, 81, 75
]

salaries = [
    51915615, 47649433, 47607350, 47607350, 47607350, 46741590, 45640165, 45640084, 45640084, 45640084,
    45183960, 43219440, 41000000, 40806300, 40600080, 40064220, 40064220, 40064220, 39270150, 37893408,
    37893408, 37037037, 36861707, 36016200, 36016200, 36016200, 36000000, 34005250, 34005250, 34005250,
    34005250, 33833400, 33833400, 33833400, 33833400, 33833400, 33833400, 32600060, 32600060, 32600060,
    32459438, 31830357, 31500000, 30800000, 30600000, 29825000, 29320988, 28600000, 28226880, 27955357,
    27586207, 27102202, 27000000, 26346666, 26346666, 25568182, 25340000, 25300000, 24460000, 24460000,
    24410000, 24410000, 24410000, 24410000, 24410000, 24410000, 24410000, 22000000, 21070000, 20975000,
    20616000, 20465117, 20357143
]


players = buildList(names, ratings, salaries)
print("-------------------------")
testMaxVal(players, 141000000)
print("-------------------------")
testGreedys(players, 141000000)

-------------------------
(BRUTE FORCE) Use search tree to allocate players within a salary cap of 141000000
Total rating of players taken = 514
    Lonzo Ball: <81, 20465117>
    Clint Capela: <81, 20616000>
    Myles Turner: <85, 20975000>
    Mikal Bridges: <83, 21070000>
    Jalen Brunson: <94, 27000000>
    Domantas Sabonis: <90, 30800000>
-------------------------
Use greedy by rating to allocate 141000000 capspace
Total rating of players taken = 293.0
    Joel Embiid: <98, 47607350>
    Nikola Jokic: <98, 47607350>
    Giannis Antetokounmpo: <97, 45640084>
Use greedy by salary to allocate 141000000 capspace
Total rating of players taken = 482.0
    Spencer Dinwiddie: <75, 20357143>
    Lonzo Ball: <81, 20465117>
    Clint Capela: <81, 20616000>
    Myles Turner: <85, 20975000>
    Mikal Bridges: <83, 21070000>
    Bruce Brown: <77, 22000000>


Explanation: Using brute force, the algorithm was able to achieve a 2k rating sum of 514 compared to the greedy algorithm (293,482). The downside of it was the runtime was noticeably slower.

There are two greedy tests. The first greedy test prioritized players with the highest ratings first and then traverses down from there. We are left with the 3 players with the highest ratings among the list. They only accumulated a total score of 293 because they are paid so much that they consume too much cap space.

The second greedy test prioritized the cheapest players from the list. They were able to accumulate a higher score than the first one because they had more players and the gap between the rating of the cheapest players and the best players is only 23,17,16 points respectively which was not enough gap to compensate for the lack of players on the first one.

Based on my observations, increasing the salary cap took so much more time for brute force to output something while the greedy algorithm took only seconds to execute. The brute force algorithm was the most successful algorithm in solving this problem as it was able to achieve the highest score among the 3. However, this came at a cost of a slow execution.

#### Conclusion:

This lesson helped me to understand the greedy and brute force algorithm, especially through the supplementary activity. I learned the pros and cons of using both methods and applying it to optimize constraints found in real life.